File: Data Structures & Algorithms/add-two-numbers/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# 5
# 7

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        curr1 = l1
        curr2 = l2
        
        carry = 0
        head = ListNode()
        res = head
        
        while curr1 or curr2:

            # If node is empty (in the case that one list is longer than the other)
            # set it to a new node with value 0
            curr1 = curr1 or ListNode(0)
            curr2 = curr2 or ListNode(0)

            sum = curr1.val + curr2.val + carry

            # For example: 15 -> 1 and 5
            carry = sum // 10
            sum = sum % 10

            res.next = ListNode(sum)

            res = res.next
            curr1 = curr1.next
            curr2 = curr2.next

        # If the sequence ends but carry is non empty
        if carry != 0:
            res.next = ListNode(carry)
        
        return head.next

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-4.cs
public class Solution {
    public ListNode MergeTwoLists(ListNode list1, ListNode list2) {
		ListNode head = new ListNode();
		
		ListNode merged = head;
		
		ListNode curr1 = list1;
		ListNode curr2 = list2;
		
		// Stop execution when one of the 2 lists are null
		while (curr1 != null && curr2 != null) {
			if (curr1.val <= curr2.val) {
				merged.next = curr1;
				curr1 = curr1.next;
			}
			else {
				merged.next = curr2;
				curr2 = curr2.next;
			}
			merged = merged.next;
		}
		
		if (curr1 != null) merged.next = curr1;
		if (curr2 != null) merged.next = curr2;
		
		return head.next;
    }
}

